### DISCUSSION TITLE
Send message only to a specific connected clients

### DISCUSSION THREAD

**Discussion by RatkoR** [enisdenjo/graphql-ws#338] [Q&A] [answered]
Is there a way to filter out a client (or multiple clients) and send message only to them?

An example would be an image with QR code: user clicks on a button, some calculations happen on server and then I'd like to send QR image to that specific client that made the click.

Each client has an UUID so I could do a filter based on that. Returning false or null in onNext() doesn't seem to help and throwing error closes connections. Is there some other way that I failed to see?

**Comment by enisdenjo** [comment, marked as answer]
This is GraphQL WebSocket library, so you may achieve that only through GraphQL. A subscription would look roughly like this:

```graphql
subscription {
  onQRAvailable(key: "unique key of client") {
    qrCodeData
  }
}
```

and them emit the event through GQL matching the key of the client.

However, a better question is: why even use WebSockets for this? Why not use HTTP which is meant for activities like the one you're describing - an HTTP request on button click whose response is the QR image data.

**Reply by RatkoR** [reply]
This was just an example (bad example).

Maybe an example with chat messages would be a better one: normally all users get all messages and this works just fine. But when admin wants to send something to some offending user, he wants that message so be delivered only to him. No need for everyone to see it. So I needed something to filter out all other users and send that one message only to him.

I managed to do it with `withFilter` from [graphql-subscriptions](https://github.com/apollographql/graphql-subscriptions#filters) package.

```javascript
import { ... } from 'graphql';

import { withFilter } from 'graphql-subscriptions';
import { DefaultType } from '....';

import { shouldSendSubsMessage } from '....';

export default {
    type: DefaultType,
    args: {
        clientId: {
            type: new GraphQLNonNull(GraphQLString)
        }
    },
    subscribe: withFilter(
            (rootValue, args, ctx, info) => ctx.redisPubSub.asyncIterator('websockets-gql'),
            (message, args, context, info) => shouldSendSubsMessage(message, args, context, info),
    )
};
```

If message has a flag so that it should be delivered to some specific user, `shouldSendSubsMessage` returns false for all those other connected users and true for user that needs to see this message.

I'd be glad to do it differently if you have a better idea.

And thanks for the great lib!

**Reply by enisdenjo** [reply]
You got it! I'd recommend doing it like that too, and that's exactly how it should be done - through GraphQL. WebSocket is just the transport layer, it shouldn't contain much logic.
